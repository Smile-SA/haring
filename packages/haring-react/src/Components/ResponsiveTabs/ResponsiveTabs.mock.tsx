import { Tabs } from '@mantine/core';

export const contentsMock = (
  <>
    <Tabs.Panel value="1">First Content</Tabs.Panel>
    <Tabs.Panel value="2">Second Content</Tabs.Panel>
    <Tabs.Panel value="3">Third Content</Tabs.Panel>
    <Tabs.Panel value="4">Fourth Content</Tabs.Panel>
    <Tabs.Panel value="5">Overflow 1 Content</Tabs.Panel>
    <Tabs.Panel value="6">Overflow 2 Content</Tabs.Panel>
  </>
);

export const tabsMock = [
  <Tabs.Tab key={1} id="tab-1" value="1">
    First
  </Tabs.Tab>,
  <Tabs.Tab key={2} data-testid="test-tab" id="tab-1" value="2">
    Second
  </Tabs.Tab>,
  <Tabs.Tab key={3} value="3">
    Third
  </Tabs.Tab>,
  <Tabs.Tab key={4} value="4">
    Fourth
  </Tabs.Tab>,
  <Tabs.Tab key={5} value="5">
    Overflow 1
  </Tabs.Tab>,
  <Tabs.Tab key={6} value="6">
    Overflow 2
  </Tabs.Tab>,
];
