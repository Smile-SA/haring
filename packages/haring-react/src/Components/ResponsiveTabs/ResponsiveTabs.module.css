.button {
  align-self: center;
  margin-bottom: 6px;
  margin-left: auto;

  &[data-expanded] {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      background-color: var(--mantine-color-gray-0);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      background-color: var(--mantine-color-dark-6);
    }
  }

  &:hover {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      background: var(--mantine-color-gray-0);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      background: var(--mantine-color-dark-6);
    }
  }
}

.container {
  position: relative;
}

.dropdown {
  border-radius: 4px;
  padding: 20px;

  /*noinspection CssInvalidAtRule*/
  @mixin light {
    background: var(--mantine-color-white);
    border-color: var(--mantine-color-white);
  }
  /*noinspection CssInvalidAtRule*/
  @mixin dark {
    background: var(--mantine-color-dark-0);
    border-color: var(--mantine-color-dark-0);
  }
}

.dropdownContainer {
  display: flex;
  flex-direction: column;

  > button {
    border: none;
    border-radius: 4px;
    padding: 10px;

    &[data-active] {
      background: var(--mantine-color-gray-0);
      border: none;
    }

    &:hover {
      background-color: var(--mantine-color-gray-0);
      border: none;
    }

    > span {
      font-weight: initial;

      /*noinspection CssInvalidAtRule*/
      @mixin light {
        color: var(--mantine-color-black);
      }
      /*noinspection CssInvalidAtRule*/
      @mixin dark {
        color: var(--mantine-color-white);
      }
    }
  }
}

.hidden {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  top: 0;
  visibility: hidden;
  width: 100%;

  & span {
    font-size: 18px;
    font-weight: 600;
  }
}

.tab {
  padding-bottom: 16px;

  &[data-active] {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      color: var(--mantine-color-black);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      color: var(--mantine-color-dark-0);
    }
  }
}

.tabs {
  flex-wrap: nowrap;
  position: absolute;
  top: 0;
  width: 100%;

  & span {
    font-size: 18px;
    font-weight: 600;
  }
}
