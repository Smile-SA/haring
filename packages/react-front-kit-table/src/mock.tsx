/* eslint-disable react-refresh/only-export-components */
export * from './Components/Table/Table.mock';
export * from './Components/TableGridView/TableGridView.mock';
