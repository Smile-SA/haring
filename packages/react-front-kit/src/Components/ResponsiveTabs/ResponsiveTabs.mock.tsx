import type { ReactNode } from 'react';

import { Tabs } from '@mantine/core';

// eslint-disable-next-line react-refresh/only-export-components
function Content({ children }: { children: ReactNode }): ReactNode {
  return <div style={{ paddingTop: 20 }}>{children}</div>;
}

export const contents = (
  <>
    <Tabs.Panel value="1">
      <Content>First Content</Content>
    </Tabs.Panel>
    <Tabs.Panel value="2">
      <Content>Second Content</Content>
    </Tabs.Panel>
    <Tabs.Panel value="3">
      <Content>Third Content</Content>
    </Tabs.Panel>
    <Tabs.Panel value="4">
      <Content>Fourth Content</Content>
    </Tabs.Panel>
    <Tabs.Panel value="5">
      <Content>Overflow 1 Content</Content>
    </Tabs.Panel>
    <Tabs.Panel value="6">
      <Content>Overflow 2 Content</Content>
    </Tabs.Panel>
  </>
);

export const tabs = [
  <Tabs.Tab key={1} value="1">
    First
  </Tabs.Tab>,
  <Tabs.Tab key={2} data-testid="test-tab" value="2">
    Second
  </Tabs.Tab>,
  <Tabs.Tab key={3} value="3">
    Third
  </Tabs.Tab>,
  <Tabs.Tab key={4} value="4">
    Fourth
  </Tabs.Tab>,
  <Tabs.Tab key={5} value="5">
    Overflow 1
  </Tabs.Tab>,
  <Tabs.Tab key={6} value="6">
    Overflow 2
  </Tabs.Tab>,
];
