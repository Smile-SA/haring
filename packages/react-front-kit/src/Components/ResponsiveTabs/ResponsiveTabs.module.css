.button {
  &[data-expanded] {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      background-color: var(--mantine-color-gray-0);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      background-color: var(--mantine-color-dark-6);
    }
  }

  &:hover {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      background: var(--mantine-color-gray-0);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      background: var(--mantine-color-dark-6);
    }
  }

  align-self: center;
  margin-bottom: 6px;
  margin-left: auto;
}

.container {
  position: relative;
}

.dropdown {
  border-radius: 4px;
  padding: 20px;

  /*noinspection CssInvalidAtRule*/
  @mixin light {
    background: var(--mantine-color-white);
    border-color: var(--mantine-color-white);
  }
  /*noinspection CssInvalidAtRule*/
  @mixin dark {
    background: var(--mantine-color-dark-0);
    border-color: var(--mantine-color-dark-0);
  }
}

.dropdownContainer {
  > .mantine-Tabs-tab {
    &[data-active] {
      background: var(--mantine-color-gray-0);
      border: none;
    }

    :hover {
      background-color: var(--mantine-color-gray-0);
      border: none;
    }

    > span {
      font-weight: initial;

      /*noinspection CssInvalidAtRule*/
      @mixin light {
        color: var(--mantine-color-black);
      }
      /*noinspection CssInvalidAtRule*/
      @mixin dark {
        color: var(--mantine-color-white);
      }
    }

    border: none;
    border-radius: 4px;
    padding: 10px;
  }

  display: flex;
  flex-direction: column;
}

.hidden {
  & .mantine-Tabs-tabLabel {
    font-size: 18px;
    font-weight: 600;
  }

  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  top: 0;
  visibility: hidden;
  width: 100%;
}

.tab {
  &[data-active] {
    /*noinspection CssInvalidAtRule*/
    @mixin light {
      color: var(--mantine-color-black);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      color: var(--mantine-color-dark-0);
    }
  }

  padding-bottom: 16px;
}

.tabs {
  & .mantine-Tabs-tabLabel {
    font-size: 18px;
    font-weight: 600;
  }

  flex-wrap: nowrap;
  position: absolute;
  top: 0;
  width: 100%;
}
