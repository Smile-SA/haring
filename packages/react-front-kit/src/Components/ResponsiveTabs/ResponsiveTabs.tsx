import type { TabsProps } from '@mantine/core';
import type { ReactElement, ReactNode } from 'react';

import { ActionIcon, Tabs } from '@mantine/core';
import { useId } from '@mantine/hooks';
import { createStyles } from '@mantine/styles';
import { CaretDoubleRight } from '@phosphor-icons/react';
import { createRef, useEffect, useState } from 'react';

import { DropdownButton } from '../DropdownButton/DropdownButton';

const useStyles = createStyles(() => ({
  dropdown: {
    display: 'flex',
    flexDirection: 'column',
  },
  tabs: {
    flexWrap: 'nowrap',
  },
}));

export interface IResponsiveTabs extends TabsProps {
  children: ReactNode;
  tabs: ReactElement[];
}

export function ResponsiveTabs(props: IResponsiveTabs): ReactNode {
  const { children, tabs, ...tabsProps } = props;
  const [overflowIndex, setOverflowIndex] = useState<number>(tabs.length);
  const tabsRef = createRef<HTMLDivElement>();
  const overflowButtonId = useId();
  const { classes } = useStyles();

  // TODO: useElementSize (https://v6.mantine.dev/hooks/use-element-size/) to re-calculate overflow on resize
  useEffect(() => {
    const parentElement = tabsRef.current;
    if (parentElement) {
      const tabElements = Array.from(
        parentElement.children,
      ) as HTMLButtonElement[];
      const overflowIndex = tabElements.findIndex((el) => {
        return (
          el.id !== overflowButtonId &&
          (el.offsetLeft - parentElement.offsetWidth + 30 >
            parentElement.offsetWidth ||
            el.offsetTop - parentElement.offsetTop > parentElement.offsetHeight)
        );
      });
      setOverflowIndex(overflowIndex);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Tabs {...tabsProps} style={{ overflow: 'hidden' }}>
      <Tabs.List ref={tabsRef} className={classes.tabs}>
        {tabs.slice(0, overflowIndex)}
        {Boolean(overflowIndex < tabs.length) && (
          <DropdownButton
            buttonComponent={
              <ActionIcon>
                <CaretDoubleRight />
              </ActionIcon>
            }
            id={overflowButtonId}
          >
            <div className={classes.dropdown}>
              {tabs.slice(overflowIndex, tabs.length)}
            </div>
          </DropdownButton>
        )}
      </Tabs.List>
      {children}
    </Tabs>
  );
}
