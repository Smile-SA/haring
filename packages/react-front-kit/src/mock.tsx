/* eslint-disable react-refresh/only-export-components */

export * from './Components/ResponsiveTabs/ResponsiveTabs.mock';
export * from './Components/SidebarMenu/SidebarMenu.mock';
export * from './Components/SwitchableView/SwitchableView.mock';
export * from './Components/Thumbnail/Thumbnail.mock';
export * from './Components/ThumbnailGrid/ThumbnailGrid.mock';
export * from './Layouts/FoldableColumnLayout/FoldableColumnLayout.mock';
