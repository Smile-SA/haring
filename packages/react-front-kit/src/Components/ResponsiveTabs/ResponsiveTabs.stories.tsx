import type { Meta, StoryObj } from '@storybook/react';
import type { ReactNode } from 'react';

import { sleep } from '@smile/react-front-kit-shared/storybook-utils';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { ResponsiveTabs as Cmp } from './ResponsiveTabs';
import { contents, tabs } from './ResponsiveTabs.mock';

const meta = {
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/ResponsiveTabs',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

function Wrapper({ children }: { children: ReactNode }): ReactNode {
  return (
    <div
      style={{
        border: '1px solid darkgrey',
        boxSizing: 'content-box',
        height: 250,
        margin: 'auto',
        overflow: 'hidden',
        padding: 30,
        resize: 'both',
        width: 300,
      }}
    >
      {children}
    </div>
  );
}

export const ResponsiveTabs: IStory = {
  args: {
    children: contents,
    defaultValue: '1',
    tabs,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await expect(canvas.getByText('First Content')).toBeVisible();
    await userEvent.click(canvas.getAllByTestId('test-tab')[1]);
    await sleep(200);
    await expect(canvas.getByText('Second Content')).toBeVisible();
  },
  render: ({ children, ...props }) => (
    <Wrapper>
      <Cmp {...props}>{children}</Cmp>
    </Wrapper>
  ),
};
