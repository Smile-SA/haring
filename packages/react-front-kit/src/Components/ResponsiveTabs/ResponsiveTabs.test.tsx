import { renderWithProviders } from '@smile/react-front-kit-shared/test-utils';

import { ResponsiveTabs } from './ResponsiveTabs';
import { contentsMock, tabsMock } from './ResponsiveTabs.mock';

describe('ResponsiveTabs', () => {
  it('matches snapshot', () => {
    const { container } = renderWithProviders(
      <ResponsiveTabs id="responsive-tabs" tabs={tabsMock}>
        {contentsMock}
      </ResponsiveTabs>,
    );
    expect(container).toMatchSnapshot();
  });
});
