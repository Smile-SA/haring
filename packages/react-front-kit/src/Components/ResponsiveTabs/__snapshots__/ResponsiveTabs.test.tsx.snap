// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveTabs matches snapshot 1`] = `
<div>
  <div
    class="mantine-Tabs-root mantine-1w1w6vj"
    id="responsive-tabs"
  >
    <div
      aria-hidden="true"
      class="mantine-cdxve1"
      hidden=""
    >
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          First
        </span>
      </button>
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        data-testid="test-tab"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          Second
        </span>
      </button>
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          Third
        </span>
      </button>
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          Fourth
        </span>
      </button>
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          Overflow 1
        </span>
      </button>
      <button
        aria-selected="false"
        class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="mantine-480qv8 mantine-Tabs-tabLabel"
        >
          Overflow 2
        </span>
      </button>
    </div>
    <div
      aria-orientation="horizontal"
      class="mantine-Tabs-tabsList mantine-1c0ih2u"
      role="tablist"
    >
      <button
        aria-controls="responsive-tabs-overflow-button-dropdown"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Overflow Button"
        class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-1i00min"
        id="responsive-tabs-overflow-button-target"
        role="tab"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M141.66,133.66l-80,80a8,8,0,0,1-11.32-11.32L124.69,128,50.34,53.66A8,8,0,0,1,61.66,42.34l80,80A8,8,0,0,1,141.66,133.66Zm80-11.32-80-80a8,8,0,0,0-11.32,11.32L204.69,128l-74.35,74.34a8,8,0,0,0,11.32,11.32l80-80A8,8,0,0,0,221.66,122.34Z"
          />
        </svg>
      </button>
      <div
        aria-labelledby="responsive-tabs-overflow-button-target"
        aria-orientation="vertical"
        class="mantine-Menu-dropdown mantine-34g9l4"
        data-position="bottom-end"
        data-testid="dropdown"
        id="responsive-tabs-overflow-button-dropdown"
        role="menu"
        style="display: none; z-index: 300; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          data-autofocus="true"
          data-menu-dropdown="true"
          style="outline: 0;"
          tabindex="-1"
        >
          <div
            class="mantine-rocuxk"
          >
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              id="responsive-tabs-tab-1"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                First
              </span>
            </button>
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              data-testid="test-tab"
              id="responsive-tabs-tab-2"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                Second
              </span>
            </button>
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              id="responsive-tabs-tab-3"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                Third
              </span>
            </button>
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              id="responsive-tabs-tab-4"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                Fourth
              </span>
            </button>
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              id="responsive-tabs-tab-5"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                Overflow 1
              </span>
            </button>
            <button
              aria-selected="false"
              class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1d12h3a"
              id="responsive-tabs-tab-6"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="mantine-480qv8 mantine-Tabs-tabLabel"
              >
                Overflow 2
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-1"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-1"
      role="tabpanel"
    >
      First Content
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-2"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-2"
      role="tabpanel"
    >
      Second Content
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-3"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-3"
      role="tabpanel"
    >
      Third Content
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-4"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-4"
      role="tabpanel"
    >
      Fourth Content
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-5"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-5"
      role="tabpanel"
    >
      Overflow 1 Content
    </div>
    <div
      aria-labelledby="responsive-tabs-tab-6"
      class="mantine-Tabs-panel mantine-18ozoug"
      id="responsive-tabs-panel-6"
      role="tabpanel"
    >
      Overflow 2 Content
    </div>
  </div>
</div>
`;
